app/assets/javascripts/discourse/app/components/discovery/filter-navigation-menu.gjs

```diff
@@ -0,0 +1,565 @@
+        this.selectedIndex = this.nothingSelected
+          ? 0
+          : (this.selectedIndex + 1) % this.filteredTips.length;
+        break;
+      case "ArrowUp":
+        event.preventDefault();
+        this.selectedIndex = this.nothingSelected
+          ? this.filteredTips.length - 1
+          : (this.selectedIndex - 1 + this.filteredTips.length) %
+            this.filteredTips.length;
+        break;
+      case "Tab":
+        event.preventDefault();
+        event.stopPropagation();
+        this.selectItem(
+          this.filteredTips[this.nothingSelected ? 0 : this.selectedIndex]
+        );
+        break;
+      case " ":
+        if (!this.dmenuInstance) {
+          this.openFilterMenu();
+        }
+        break;
+      case "Enter":
+        if (this.selectedIndex >= 0) {
+          event.preventDefault();
+          event.stopPropagation();
+          this.selectItem(this.filteredTips[this.selectedIndex]);
+        } else {
+          if (!this.dmenuInstance) {
+            this.args.onChange(this.currentInputValue, true);
+            return;
+          }
+
+          this.dmenuInstance?.close().then(() => {
+            this.args.onChange(this.currentInputValue, true);
+          });
+        }
+        break;
+      case "Escape":
+        this.dmenuInstance?.close();
+        break;
+    }
+  }
+
+  async #performFilterSuggestionSearch(filterName, valueText, tip, prefix) {
+    const type = tip.type;
+    let lastTerm = valueText;
+    let results = [];
+    let prevTerms = "";
+    let splitTerms;
+
+    if (tip.delimiters) {
+      const delimiters = tip.delimiters.map((s) => s.name);
+      splitTerms = lastTerm.split(new RegExp(`[${delimiters.join("")}]`));
+      lastTerm = splitTerms[splitTerms.length - 1];
+      prevTerms =
+        lastTerm === "" ? valueText : valueText.slice(0, -lastTerm.length);
+    }
+
+    lastTerm = (lastTerm || "").toLowerCase().trim();
+
+    switch (type) {
+      case "tag":
+        results = await this.filterSuggestions.getTagSuggestions(
+          prefix,
+          filterName,
+          prevTerms,
+          lastTerm
+        );
+        break;
+
+      case "category":
+        results = this.filterSuggestions.getCategorySuggestions(
+          prefix,
+          filterName,
+          prevTerms,
+          lastTerm
+        );
+        break;
+
+      case "username":
+        results = await this.filterSuggestions.getUserSuggestions(
+          prefix,
+          filterName,
+          prevTerms,
+          lastTerm
+        );
+        break;
+
+      case "date":
+        results = this.filterSuggestions.getDateSuggestions(
+          prefix,
+          filterName,
+          prevTerms,
+          lastTerm
+        );
+        break;
+
+      case "number":
+        results = this.filterSuggestions.getNumberSuggestions(
+          prefix,
+          filterName,
+          prevTerms,
+          lastTerm
+        );
+        break;
+    }
```

Comment:
I think we can omit a lot of the duplication here, given all the params are the same. also worth nothing... category needs an await and slight re-implementation for category scaling. I wonder if for evenness of API all of these get an await.